### PR TITLE
Added consumer token use to moodycamel wrapper

### PR DESCRIPTION
**Pull request by cameron314**

### CODE REVIEW

File: include/atomic_queue/moodycamel.h [hunk 1]
```diff
@@ -15,29 +15,53 @@ namespace atomic_queue {
...
+    MoodyCamelQueue()
+        : moodycamel::ConcurrentQueue<T>(Capacity) {
+        ctok() = moodycamel::ConcurrentQueue<T>::consumer_token_t(*this);  // update instance
+    }
 
     void push(T element) {
-        while(!this->try_enqueue(element))
+        // TODO: To be able to use try_enqueue properly, particularly with tokens,
+        // requires the number of producer and consumer threads to be passed to the
+        // constructor
```
Comment (max0x7ba) [inline include/atomic_queue/moodycamel.h:28]: Would you like to pass that number into the constructor then?

Comment (cameron314) [reply include/atomic_queue/moodycamel.h:28]: That would be ideal but I didn't want to change your interface/API.

**Comment by cameron314** [comment]
Note that between consumer and producer tokens, the consumer ones are most important for performance.

File: include/atomic_queue/moodycamel.h (hunk 1 above)
Comment (max0x7ba) [reply include/atomic_queue/moodycamel.h:28]: Feel free to change the implementation and the API of the benchmark, please.

If not, I will try to route elegantly the consumer/producer number into your queue constructor by the end of this week.

Comment (cameron314) [reply include/atomic_queue/moodycamel.h]: Alright I've pushed a change that passes the producer thread count and adds tokens. I changed the moodycamel, Boost, and Xenium adapters, but AtomicQueuea and TBB probably need new adapters.

I can't compile on my platform (Windows with no Boost, TBB, etc.) so my changes may need a bit of work before merging.

**Comment by max0x7ba** [comment]
Hi Cameron,

Many thanks for your improvements.

I perused your changes and managed to make a complete implementation that: 

1. Passes the number of producers and consumers into `moodycamel::ConcurrentQueue` constructor.
2. Uses thread-specific producer/consumer tokens as extra arguments for push/pop of `moodycamel::ConcurrentQueue`.

I copied your queue-to-benchmark adaptors, but implemented another argument injection mechanism that required no changes to other queue/adaptors. 

Changes are in commit https://github.com/max0x7ba/atomic_queue/commit/416905671ce7a8caa7d7fb238c9da7d9624e1eb3

If you don't have any comments, improvements or concerns, I will re-run the benchmarks and rebuild the charts.

Maxim

**Comment by cameron314** [comment]
Nice! Looks good to me. Thanks for taking the time for this, I know my API is a bit unconventional.
Looking forward to the benchmark results :-)

**Comment by max0x7ba** [comment]
Hi Cameron,

You may be pleased to know that these changes improved the performance of your queues in the benchmarks considerably.

Before:
```
        moodycamel::ConcurrentQueue, 1,i:  23,189,568 msg/sec (mean:  21,191,257 stdev:   1,040,230)
        moodycamel::ConcurrentQueue, 1,s:  22,856,810 msg/sec (mean:  20,837,913 stdev:     766,683)
        moodycamel::ConcurrentQueue, 2,i:  14,920,987 msg/sec (mean:  12,800,646 stdev:     522,067)
        moodycamel::ConcurrentQueue, 2,s:  15,080,986 msg/sec (mean:  12,702,508 stdev:     501,296)
        moodycamel::ConcurrentQueue, 3,i:  13,841,999 msg/sec (mean:  12,606,367 stdev:     263,036)
        moodycamel::ConcurrentQueue, 3,s:  13,951,343 msg/sec (mean:  12,459,394 stdev:     258,860)
        moodycamel::ConcurrentQueue, 4,i:  14,859,935 msg/sec (mean:  13,335,839 stdev:     708,284)
        moodycamel::ConcurrentQueue, 4,s:  14,923,133 msg/sec (mean:  13,141,623 stdev:     757,457)
        moodycamel::ConcurrentQueue, 5,i:  19,016,106 msg/sec (mean:  15,059,781 stdev:   1,041,940)
        moodycamel::ConcurrentQueue, 5,s:  16,380,068 msg/sec (mean:  14,467,530 stdev:     540,038)
        moodycamel::ConcurrentQueue, 6,i:  19,893,649 msg/sec (mean:  16,290,913 stdev:   1,085,651)
        moodycamel::ConcurrentQueue, 6,s:  17,085,918 msg/sec (mean:  15,014,489 stdev:     580,440)
        moodycamel::ConcurrentQueue, 7,i:  20,708,687 msg/sec (mean:  17,123,425 stdev:   1,024,438)
        moodycamel::ConcurrentQueue, 7,s:  17,502,870 msg/sec (mean:  15,328,828 stdev:     491,546)
        moodycamel::ConcurrentQueue, 8,i:  20,938,407 msg/sec (mean:  17,810,739 stdev:     902,165)
        moodycamel::ConcurrentQueue, 8,s:  17,123,559 msg/sec (mean:  15,610,513 stdev:     432,508)
      moodycamel::ReaderWriterQueue, 1,i: 295,716,975 msg/sec (mean:  60,958,370 stdev:  24,591,233)
      moodycamel::ReaderWriterQueue, 1,s: 278,924,720 msg/sec (mean:  72,090,845 stdev:  38,103,719)
```

After:
```
        moodycamel::ConcurrentQueue, 1,i:  26,851,462 msg/sec (mean:  24,444,363 stdev:   1,768,641)
        moodycamel::ConcurrentQueue, 1,s:  29,453,011 msg/sec (mean:  25,527,119 stdev:   1,174,150)
        moodycamel::ConcurrentQueue, 2,i:  52,431,952 msg/sec (mean:  16,505,388 stdev:   4,634,975)
        moodycamel::ConcurrentQueue, 2,s:  48,675,054 msg/sec (mean:  16,016,698 stdev:   3,421,829)
        moodycamel::ConcurrentQueue, 3,i:  43,648,056 msg/sec (mean:  17,180,551 stdev:   4,264,074)
        moodycamel::ConcurrentQueue, 3,s:  43,070,009 msg/sec (mean:  16,912,155 stdev:   4,020,687)
        moodycamel::ConcurrentQueue, 4,i:  45,476,352 msg/sec (mean:  17,895,106 stdev:   4,437,429)
        moodycamel::ConcurrentQueue, 4,s:  45,837,232 msg/sec (mean:  18,562,667 stdev:   5,718,200)
        moodycamel::ConcurrentQueue, 5,i:  47,121,047 msg/sec (mean:  19,182,942 stdev:   3,459,405)
        moodycamel::ConcurrentQueue, 5,s:  47,841,058 msg/sec (mean:  19,308,157 stdev:   6,781,750)
        moodycamel::ConcurrentQueue, 6,i:  63,384,866 msg/sec (mean:  21,432,650 stdev:   6,611,307)
        moodycamel::ConcurrentQueue, 6,s:  46,951,462 msg/sec (mean:  18,940,226 stdev:   5,894,943)
        moodycamel::ConcurrentQueue, 7,i:  66,347,129 msg/sec (mean:  22,544,800 stdev:   7,439,106)
        moodycamel::ConcurrentQueue, 7,s:  49,691,434 msg/sec (mean:  19,531,464 stdev:   6,005,440)
        moodycamel::ConcurrentQueue, 8,i:  64,375,322 msg/sec (mean:  22,827,986 stdev:   5,980,747)
        moodycamel::ConcurrentQueue, 8,s:  61,187,363 msg/sec (mean:  21,936,042 stdev:  10,065,222)
      moodycamel::ReaderWriterQueue, 1,i: 536,819,654 msg/sec (mean: 217,131,994 stdev: 185,262,413)
      moodycamel::ReaderWriterQueue, 1,s: 538,733,018 msg/sec (mean: 295,875,273 stdev: 177,103,961)
```

**Pull request closed without merging**